crates/circuit/src/dag_circuit.rs

```diff
@@ -2571,78 +2571,99 @@ def _format(operand):
-        //         new_wires.update(op.target)
-        //     else:
-        //         new_wires.update(node_resources(op.target).clbits)
-        //
-        // if propagate_condition and not (
-        //     isinstance(node.op, ControlFlowOp) or isinstance(op, ControlFlowOp)
-        // ):
-        //     if new_condition is not None:
-        //         raise DAGCircuitError(
-        //             "Cannot propagate a condition to an operation that already has one."
-        //         )
-        //     if (old_condition := getattr(node.op, "condition", None)) is not None:
-        //         if not isinstance(op, Instruction):
-        //             raise DAGCircuitError("Cannot add a condition on a generic Operation.")
-        //         if not isinstance(node.op, ControlFlowOp):
-        //             op = op.c_if(*old_condition)
-        //         else:
-        //             op.condition = old_condition
-        //         new_wires.update(condition_resources(old_condition).clbits)
-        //
-        // if new_wires != current_wires:
-        //     # The new wires must be a non-strict subset of the current wires; if they add new wires,
-        //     # we'd not know where to cut the existing wire to insert the new dependency.
-        //     raise DAGCircuitError(
-        //         f"New operation '{op}' does not span the same wires as the old node '{node}'."
-        //         f" New wires: {new_wires}, old wires: {current_wires}."
-        //     )
-        //
-        // if inplace:
-        //     if op.name != node.op.name:
-        //         self._increment_op(op)
-        //         self._decrement_op(node.op)
-        //     node.op = op
-        //     return node
-        //
-        // new_node = copy.copy(node)
-        // new_node.op = op
-        // self._multi_graph[node._node_id] = new_node
-        // if op.name != node.op.name:
-        //     self._increment_op(op)
-        //     self._decrement_op(node.op)
-        // return new_node
-        todo!()
+    ) -> PyResult<Py<PyAny>> {
+        let py = op.py();
+        // Extract information from node that is going to be replaced
+        let old_packed = match self.dag.node_weight(node.as_ref().node.unwrap()) {
+            Some(NodeType::Operation(old_packed)) => old_packed.clone(),
+            Some(_) => {
+                return Err(DAGCircuitError::new_err(
+                    "'node' must be of type 'DAGOpNode'.",
+                ))
+            }
+            None => return Err(DAGCircuitError::new_err("'node' not found in DAG.")),
+        };
+        // Extract information from new op
+        let new_op = convert_py_to_operation_type(py, op.clone().unbind())?;
+
+        // If either operation is a control-flow operation, propagate_condition is ignored
+        let new_condition = match propagate_condition
+            && !new_op.operation.control_flow()
+            && !node.instruction.operation.control_flow()
```

Comment:
I think it'd be best to only use `node` to get the ID of the node the user is referring to, and otherwise use `old_packed` to read any information. This way, the data in `self.dag` is the single source of truth (i.e. this way, the user can't get a weird outcome by crafting a `node` with an ID and payload that don't match what is in our graph).